src/plugins/intel_npu/src/plugin/npuw/util_xarch.cpp

```diff
@@ -1473,3 +1514,160 @@ void ov::npuw::util::XARCH::copy_row_as_column(const ov::SoPtr<ov::ITensor>& fro
     from->copy_to(to._ptr);
 #endif
 }
+
+void ov::npuw::util::XARCH::transpose_i4(const uint8_t* src, uint8_t* dst, size_t rows, size_t cols) {
+#if defined(HAVE_AVX2)
+    size_t c_step = 8;
+    size_t r_step = 8;
+
+    size_t cols32 = cols / c_step;
+    size_t rows32 = rows / r_step;
+
+    // lambda: get int32 pointer for each row of src.
+    auto get_src_int32 = [&](size_t i) {
+        return reinterpret_cast<const int32_t*>(src + (i * cols / 2));
+    };
+    // lambda: get int32 pointer for each column of dst.
+    auto get_dst_int32 = [&](size_t j) {
+        return reinterpret_cast<int32_t*>(dst + (j * rows / 2));
+    };
+
+    // Main block processing.
+    for (size_t c = 0; c < cols32; ++c) {
+        for (size_t r = 0; r < rows32; ++r) {
+            // For each row of src, recalculate the pointer.
+            const int32_t* src_row0 = get_src_int32(r * r_step + 0);
```

Comment:
I think we can avoid this:

```c++
        auto get_src_int32 = [&](size_t r, size_t c, size_t i) {
            return reinterpret_cast<const int32_t*>(&src[(r*r_step + i)*(cols/2)])[c];
        };

...

                __m256i column = _mm256_set_epi32(
                    get_src_int32(r, c, 7),
                    get_src_int32(r, c, 6),
                    get_src_int32(r, c, 5),
                    get_src_int32(r, c, 4),
                    get_src_int32(r, c, 3),
                    get_src_int32(r, c, 2),
                    get_src_int32(r, c, 1),
                    get_src_int32(r, c, 0)
                );
```

This way it is more compact

Reply:
Fixed it the latest [commit](https://github.com/openvinotoolkit/openvino/pull/30899/commits/65d4ca07cfe6663cf4e169d57c8b91662e374aff).